### ISSUE TITLE
ext() returns 0,0,1,1  

### ISSUE THREAD

**Issue by micha-silver** [rspatial/terra#409]
I am running the `supercells` function from the [supercells package](https://github.com/Nowosad/supercells). That function makes use of terra::ext() to assign the correct extent to a newly created matrix of superpixels. I have run the function repeatedly on two different Debian machines, and at random I get the result located between (0,0) and (1,1) instead of the correct coordinates. Here's an image of correct results and a "failed" run:
 
![Two_supercell_runs](https://user-images.githubusercontent.com/3841768/142735350-f5ea83d2-0dd3-4c19-93dc-eac248563bcc.png)

The left image shows properly located superpixels. The right image show a layer of superpixels wrongly placed (note the scale and coord grid). The images were created from separate runs of a test script, performed one after the other. 

**Additional information:** 

A colleague running Windows does not have this problem. And Jakub, maintainer of `supercells` has done extensive testing (see the [issue #17](https://github.com/Nowosad/supercells/issues/17) ) on Fedora and Redhat, and also cannot replicate the problem.  So this seems to be Debian specific. :-(

My sessionInfo():
```
> sessionInfo()
R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 11 (bullseye)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] supercells_0.1.2 future_1.23.0    sf_1.0-4         terra_1.4-21    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7             parallelly_1.28.1      magrittr_2.0.1        
 [4] units_0.7-2            tidyselect_1.1.1       R6_2.5.1              
 [7] rlang_0.4.12           fansi_0.5.0            dplyr_1.0.7           
[10] globals_0.14.0         tools_4.1.2            parallel_4.1.2        
[13] grid_4.1.2             KernSmooth_2.23-20     utf8_1.2.2            
[16] e1071_1.7-9            DBI_1.1.1              ellipsis_0.3.2        
[19] class_7.3-19           digest_0.6.28          assertthat_0.2.1      
[22] tibble_3.1.6           lifecycle_1.0.1        crayon_1.4.2          
[25] philentropy_0.6.0.9000 purrr_0.3.4            vctrs_0.3.8           
[28] codetools_0.2-18       glue_1.5.0             proxy_0.4-26          
[31] compiler_4.1.2         pillar_1.6.4           generics_0.1.1        
[34] classInt_0.4-3         listenv_0.8.0          pkgconfig_2.0.3    
```

The procedure in the `supercells()` [(code on github)](https://github.com/micha-silver/supercells/blob/master/R/supercells.R) function is as follows:

1. Save the SpatRaster extent
2. Prepare a matrix of values from the SpatRaster, at it's dim()
3. Call an Rpp procedure to add an id to each matrix cell identifying the superpixels and return that matrix.
4. Read matrix with `rast()` to revert to SpatRaster, and use terra::ext() to apply the saved extent to the matrix
5. Convert to `sf`

But sometimes step 4 fails and leaves the SpatRaster at (0,0,1,1), apparently randomly! Any ideas what might be causing this erratic behavior are welcome.

**Comment by rhijmans** [comment]
Apparently this happens when raster data is written to a temp file. I hope that should help Jakub @Nowosad in figuring out what is going on. I am closing this issue here, because for the time being it is a supercells problem (and since I commented there, I will receive updates from supercells issue [#17](https://github.com/Nowosad/supercells/issues/17)).

**Comment by micha-silver** [comment]
@rhijmans 
Thanks for the rapid response!
I'll also keep in touch with @Nowosad to see how this unfolds.

**Comment by Nowosad** [comment]
I was able to simplify this issue. Long story short - it seems that `terra::as.polygons()` drops extent in certain situations: 


``` r
library(terra)
#> terra version 1.4.18
vol = rast(system.file("raster/volcano.tif", package = "supercells"))
ext_x = ext(vol)

slic_sf = terra::rast(matrix(1:100, ncol = 10))
terra::ext(slic_sf) = ext_x
slic_sf
#> class       : SpatRaster 
#> dimensions  : 10, 10, 1  (nrow, ncol, nlyr)
#> resolution  : 61, 87  (x, y)
#> extent      : 2667400, 2668010, 6478705, 6479575  (xmin, xmax, ymin, ymax)
#> coord. ref. :  
#> source      : memory 
#> name        : lyr.1 
#> min value   :     1 
#> max value   :   100
terra::as.polygons(slic_sf, dissolve = TRUE)
#>  class       : SpatVector 
#>  geometry    : polygons 
#>  dimensions  : 100, 1  (geometries, attributes)
#>  extent      : 2667400, 2668010, 6478705, 6479575  (xmin, xmax, ymin, ymax)
#>  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
#>  names       : lyr.1
#>  type        : <num>
#>  values      :     1
#>                    2
#>                    3
terraOptions(todisk = TRUE)
slic_sf = terra::rast(matrix(1:100, ncol = 10))
terra::ext(slic_sf) = ext_x
slic_sf
#> class       : SpatRaster 
#> dimensions  : 10, 10, 1  (nrow, ncol, nlyr)
#> resolution  : 61, 87  (x, y)
#> extent      : 2667400, 2668010, 6478705, 6479575  (xmin, xmax, ymin, ymax)
#> coord. ref. :  
#> source      : spat_xH9HrLwGHl0yi2y.tif 
#> name        : lyr.1 
#> min value   :     1 
#> max value   :   100
terra::as.polygons(slic_sf, dissolve = TRUE)
#>  class       : SpatVector 
#>  geometry    : polygons 
#>  dimensions  : 100, 1  (geometries, attributes)
#>  extent      : 0, 1, 0, 1  (xmin, xmax, ymin, ymax)
#>  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
#>  names       : lyr.1
#>  type        : <num>
#>  values      :     1
#>                    2
#>                    3
```

<sup>Created on 2021-11-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by micha-silver** [comment]
And this happens only when the SpatRaster is saved to `tempdir()` ??

**Comment by rhijmans** [comment]
I now get 

```
library(terra)
vol = rast(system.file("raster/volcano.tif", package = "supercells"))
e = ext(vol)

terraOptions(todisk = FALSE)
r = rast(matrix(1:100, ncol = 10))
ext(r) = e
as.polygons(r, dissolve = TRUE) |> ext()
#SpatExtent : 2667400, 2668010, 6478705, 6479575 (xmin, xmax, ymin, ymax)

terraOptions(todisk = TRUE)
p <- as.polygons(r, dissolve = TRUE)
ext(p)
#SpatExtent : 2667400, 2668010, 6478705, 6479575 (xmin, xmax, ymin, ymax)
crs(p)
#[1] ""
``` 

The error occurred when the input SpatRaster is in memory, but the "mask" (a layer derived from the input that is 0 for cells that should be excluded; only used when `na.rm=TRUE`) is from disk. This is not a typical situation but could happen if the input raster is large and has its cells values in RAM, but relatively large. I am not sure why this happens and if this is a GDAL bug or what, but for now it seems that things at least work as expected.

**Comment by rhijmans** [comment]
I believe it is fixed now. 

```
library(terra)
vol = rast(system.file("raster/volcano.tif", package = "supercells"))
e = ext(vol)

terraOptions(todisk = FALSE)
r = rast(matrix(1:100, ncol = 10))
ext(r) = e
as.polygons(r, dissolve = TRUE) |> ext()
#SpatExtent : 2667400, 2668010, 6478705, 6479575 (xmin, xmax, ymin, ymax)

terraOptions(todisk = TRUE)
r = rast(matrix(1:100, ncol = 10))
ext(r) = e
as.polygons(r, dissolve = TRUE) |> ext()
#SpatExtent : 2667400, 2668010, 6478705, 6479575 (xmin, xmax, ymin, ymax)
```

The underlying reason was that this:

```
terraOptions(todisk = TRUE)
r = rast(matrix(1:100, ncol = 10))
ext(r) = e
```
Creates a raster on disk with default extent c(0,1,0,1), but the object has the new extent. When terra processes data  it always takes the extent from the object. But here we use an algorithm that comes with GDAL and goes directly to the data source on disk with the original extent. For now, it is fixed by avoiding that `r = rast(matrix(1:100, ncol = 10))` write a temp file in the first place. You can now also use the more compact `r = rast(matrix(1:100, ncol = 10), ext=e)`

But I do need to work on a more general fix, probably by having a flag that indicates if an extent was changed, to avoid these problems with GDAL algorithms, illustrated here:

```
f <- system.file("ex/elev.tif", package="terra")
r <- rast(f)
ext(r) <- c(0,1,0,1)
as.polygons(r, dissolve = TRUE) |> ext()
SpatExtent : 5.75, 6.525, 49.45, 50.1833333333333 (xmin, xmax, ymin, ymax)
``` 
(the original extent is used for the polygons; which should not be the case)

**Comment by rhijmans** [comment]
I believe the remaining issue has now also been solved

```
library(terra)
f <- system.file("ex/elev.tif", package="terra")
r <- rast(f)
ext(r) <- c(1,2,1,2)
as.polygons(r) |> ext()
#SpatExtent : 1.01052631578947, 1.98947368421053, 1.01111111111111, 1.98888888888889 (xmin, xmax, ymin, ymax)
as.polygons(r, na.rm=FALSE) |> ext()
#SpatExtent : 1, 2, 1, 2 (xmin, xmax, ymin, ymax)
```

**Comment by micha-silver** [comment]
After updating `terra` (as well as `supercells`) I ran my test script on my data file for four cycles, so producing 4 `sf` polygon objects. I used the `supercells` "chunks" option to break the original raster into 2500x2500 chunks. Of the results, three had the correct extent, and one "left" a chunk at (0,0,1,1).
I'll run again on a separate computer to verify. Is there anything else I need to update?
Thanks, Micha
